*Notebook created by: Thabor Walbeek, March 2020 for learning purposes*

# Covid-19

![](covid.jpg)

Coronavirus disease (COVID-19) is an infectious disease caused by a newly discovered coronavirus.

Most people infected with the COVID-19 virus will experience mild to moderate respiratory illness and recover without requiring special treatment.  Older people, and those with underlying medical problems like cardiovascular disease, diabetes, chronic respiratory disease, and cancer are more likely to develop serious illness.

The best way to prevent and slow down transmission is be well informed about the COVID-19 virus, the disease it causes and how it spreads. Protect yourself and others from infection by washing your hands or using an alcohol based rub frequently and not touching your face. 

The COVID-19 virus spreads primarily through droplets of saliva or discharge from the nose when an infected person coughs or sneezes, so it’s important that you also practice respiratory etiquette (for example, by coughing into a flexed elbow).

At this time, there are no specific vaccines or treatments for COVID-19. However, there are many ongoing clinical trials evaluating potential treatments. WHO will continue to provide updated information as soon as clinical findings become available.

https://www.who.int/health-topics/coronavirus#tab=tab_1


## 0. Analysis Questions to be answered

1. How many rows does the data set contain?
2. How many countries are mentioned in the data set?
3. Explain the definition of 'lat' and 'long', what do they represent?
4. Which column in the data set contains missing values and explain why?
5. Why is it important to always double-check your outcome of values?



## 1. Exploratory Data Analysis

This notebook describes several steps for beginners to explore the 3 data sets for the COVID-19 virus. The 3 data sets are available on Kaggle (www.kaggle.com)

- time_series_covid19_confirmed_global.csv
- time_series_covid19_deaths_global.csv
- time_series_covid19_recovered_global.csv

Data set taken on 28th March 2020.

For the most accurate data please look at:

- https://www.worldometers.info/
- https://experience.arcgis.com/experience/685d0ace521648f8a5beeeee1b9125cd (WHO)

### 1.1 Load the data sets

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns                       #visualisation
import matplotlib.pyplot as plt             #visualisation

In [2]:
confirmed = pd.read_csv("https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_deaths_global.csv")
deaths = pd.read_csv("https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_deaths_global.csv")
recovered = pd.read_csv("https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_recovered_global.csv")

By loading the data sets, we have the data available to explore. In the first lines of code we **import** some packages, that will help in using pre-defined functions.

We will first have a look at how the data sets look like, to understand what information is available to us:

### 1.2 First glance on data

In [3]:
confirmed.head(10)

Province/State       Country/Region      Lat      Long  \
0                           NaN          Afghanistan  33.0000   65.0000   
1                           NaN              Albania  41.1533   20.1683   
2                           NaN              Algeria  28.0339    1.6596   
3                           NaN              Andorra  42.5063    1.5218   
4                           NaN               Angola -11.2027   17.8739   
5                           NaN  Antigua and Barbuda  17.0608  -61.7964   
6                           NaN            Argentina -38.4161  -63.6167   
7                           NaN              Armenia  40.0691   45.0382   
8  Australian Capital Territory            Australia -35.4735  149.0124   
9               New South Wales            Australia -33.8688  151.2093   

   1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20   ...     3/18/20  \
0        0        0        0        0        0        0   ...          22   
1        0        0        0        0        0        0   ...          59   
2        0        0        0        0        0        0   ...          74   
3        0        0        0        0        0        0   ...          39   
4        0        0        0        0        0        0   ...           0   
5        0        0        0        0        0        0   ...           1   
6        0        0        0        0        0        0   ...          79   
7        0        0        0        0        0        0   ...          84   
8        0        0        0        0        0        0   ...           3   
9        0        0        0        0        3        4   ...         267   

   3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  3/26/20  \
0       22       24       24       40       40       74       84       94   
1       64       70       76       89      104      123      146      174   
2       87       90      139      201      230      264      302      367   
3       53       75       88      113      133      164      188      224   
4        0        1        2        2        3        3        3        4   
5        1        1        1        1        3        3        3        7   
6       97      128      158      266      301      387      387      502   
7      115      136      160      194      235      249      265      290   
8        4        6        9       19       32       39       39       53   
9      307      353      436      669      669      818     1029     1219   

   3/27/20  
0      110  
1      186  
2      409  
3      267  
4        4  
5        7  
6      589  
7      329  
8       62  
9     1405  

[10 rows x 70 columns]

From this first look at the data we can see that we have several rows and several columns. To understand this, let's describe the columns and what value they contain:

### 1.2 Data Dictionary

| Column Name | Description |
| -- | -- |
| Province/State | Some countries have specific information about provinces or states, and hence have several rows for each of them |
| Country/Region | The name of the country |
| Lat | the value of the latitude of the country/region and/or the specific province/state |
| Long | the value of the longitude of the country/region and/or the specific province/state |
| 1/22/20 | The first available date with information on confirmed cases |
| ... | ... |
| 03/27/20 | The last available date with information on confirmed cases |

There are 70 columns in total with above information per column. Let's check the number of rows:

### 1.3 Shape of the data

In [4]:
confirmed.shape

(249, 70)

There are in total 249 rows in the data set (including the header row). As some countries have multiple rows (specified in provinces/state), we want to check the number of unique countries. For that we can do the following:

In [5]:
confirmed['Country/Region'].nunique()

176

So from 248 rows in the data set for confirmed cases we have 176 unique countries.

In [6]:
confirmed.tail()

Province/State         Country/Region        Lat        Long  \
244                    NaN  Saint Kitts and Nevis  17.357822  -62.782998   
245  Northwest Territories                 Canada  64.825500 -124.845700   
246                  Yukon                 Canada  64.282300 -135.000000   
247                    NaN                 Kosovo  42.602636   20.902977   
248                    NaN                  Burma  21.916200   95.956000   

     1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20   ...     3/18/20  \
244        0        0        0        0        0        0   ...           0   
245        0        0        0        0        0        0   ...           0   
246        0        0        0        0        0        0   ...           0   
247        0        0        0        0        0        0   ...           0   
248        0        0        0        0        0        0   ...           0   

     3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  3/26/20  \
244        0        0        0        0        0        0        2        2   
245        0        0        0        0        0        0        0        1   
246        0        0        0        0        0        0        0        3   
247        0        0        0        0        0        0        0       71   
248        0        0        0        0        0        0        0        0   

     3/27/20  
244        2  
245        1  
246        3  
247       86  
248        8  

[5 rows x 70 columns]

### 1.4 Checking Missing Values

Next up we want to know how many missing values we have in the data set. If, for example, we see some rows with empty or missing values, we can remove them to be able to report in a better way.

In [7]:
print(confirmed.isnull().sum())

Province/State    173
Country/Region      0
Lat                 0
Long                0
1/22/20             0
1/23/20             0
1/24/20             0
1/25/20             0
1/26/20             0
1/27/20             0
1/28/20             0
1/29/20             0
1/30/20             0
1/31/20             0
2/1/20              0
2/2/20              0
2/3/20              0
2/4/20              0
2/5/20              0
2/6/20              0
2/7/20              0
2/8/20              0
2/9/20              0
2/10/20             0
2/11/20             0
2/12/20             0
2/13/20             0
2/14/20             0
2/15/20             0
2/16/20             0
                 ... 
2/27/20             0
2/28/20             0
2/29/20             0
3/1/20              0
3/2/20              0
3/3/20              0
3/4/20              0
3/5/20              0
3/6/20              0
3/7/20              0
3/8/20              0
3/9/20              0
3/10/20             0
3/11/20             0
3/12/20   

From here we can see that all columns are not having any missing values, except for the first column 'Province/State'.

This is because most countries are not submitting details on the detailed level of provinces or states. So it might be better to aggregate those values country-wise total. From the available data set let's have a look at the aggregated values per country:

In [8]:
# Create a data set to unpivot (remove columns for 'Lat' and 'Long')

confirmed_unpivot = confirmed
confirmed_unpivot = confirmed_unpivot.drop(['Lat','Long'], axis=1)

In [9]:
# Unpivot this data set and group them by country to get the sum of cases per day per country

confirmed_melt = confirmed_unpivot.melt(id_vars=['Province/State', 'Country/Region'], 
                                        var_name='Date', 
                                        value_name='Confirmed Cases')
confirmed_melt.groupby(['Country/Region']).sum()

Confirmed Cases
Country/Region                         
Afghanistan                         651
Albania                            1357
Algeria                            2563
Andorra                            1399
Angola                               22
Antigua and Barbuda                  33
Argentina                          3220
Armenia                            2254
Australia                         19604
Austria                           47964
Azerbaijan                          966
Bahamas                              50
Bahrain                            5875
Bangladesh                          350
Barbados                            129
Belarus                             975
Belgium                           44391
Belize                                8
Benin                                41
Bhutan                               31
Bolivia                             323
Bosnia and Herzegovina             1457
Brazil                            18676
Brunei                             1186
Bulgaria                           2141
Burkina Faso                        965
Burma                                 8
Cabo Verde                           26
Cambodia                            841
Cameroon                            505
...                                 ...
Slovenia                           5637
Somalia                              15
South Africa                       5027
Spain                            405473
Sri Lanka                          1010
Sudan                                29
Suriname                             55
Sweden                            28311
Switzerland                       93705
Syria                                18
Taiwan*                            3336
Tanzania                             97
Thailand                           8698
Timor-Leste                           6
Togo                                164
Trinidad and Tobago                 436
Tunisia                            1180
Turkey                            17799
US                               474625
Uganda                               71
Ukraine                            1040
United Arab Emirates               3482
United Kingdom                    80715
Uruguay                            1504
Uzbekistan                          493
Venezuela                           788
Vietnam                            2153
West Bank and Gaza                  913
Zambia                               67
Zimbabwe                             24

[176 rows x 1 columns]

Looking at this data we can spot something weird, which might indicate that the action we performed is not done right. Have a look at the value for US, it contains a value that is higher than the actual confirmed cases in total for all countries. Let's check this, as we have to make sure the data is correct.

In [10]:
confirmed_US = confirmed[confirmed['Country/Region']=='US']
confirmed_US

Province/State Country/Region      Lat     Long  1/22/20  1/23/20  \
225            NaN             US  37.0902 -95.7129        1        1   

     1/24/20  1/25/20  1/26/20  1/27/20   ...     3/18/20  3/19/20  3/20/20  \
225        2        2        5        5   ...        7783    13677    19100   

     3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  3/26/20  3/27/20  
225    25489    33276    43847    53740    65778    83836   101657  

[1 rows x 70 columns]

If we have a closer look at the 'US' values, we see that the values are already aggregating the data per day, where new values are added to the previous values. So for this let's first take the maximum value of each row in the original source, and then aggregate them on country level.

In [11]:
confirmed_melt = confirmed_unpivot.melt(id_vars=['Province/State', 'Country/Region'], 
                                        var_name='Date', 
                                        value_name='Confirmed Cases')
confirmed_melt.groupby(['Province/State','Country/Region']).max()
confirmed_new = confirmed_melt.groupby(['Country/Region']).max()
confirmed_new

Date  Confirmed Cases
Country/Region                                 
Afghanistan             3/9/20              110
Albania                 3/9/20              186
Algeria                 3/9/20              409
Andorra                 3/9/20              267
Angola                  3/9/20                4
Antigua and Barbuda     3/9/20                7
Argentina               3/9/20              589
Armenia                 3/9/20              329
Australia               3/9/20             1405
Austria                 3/9/20             7657
Azerbaijan              3/9/20              165
Bahamas                 3/9/20               10
Bahrain                 3/9/20              466
Bangladesh              3/9/20               48
Barbados                3/9/20               24
Belarus                 3/9/20               94
Belgium                 3/9/20             7284
Belize                  3/9/20                2
Benin                   3/9/20                6
Bhutan                  3/9/20                3
Bolivia                 3/9/20               61
Bosnia and Herzegovina  3/9/20              237
Brazil                  3/9/20             3417
Brunei                  3/9/20              115
Bulgaria                3/9/20              293
Burkina Faso            3/9/20              180
Burma                   3/9/20                8
Cabo Verde              3/9/20                5
Cambodia                3/9/20               99
Cameroon                3/9/20               91
...                        ...              ...
Slovenia                3/9/20              632
Somalia                 3/9/20                3
South Africa            3/9/20             1170
Spain                   3/9/20            65719
Sri Lanka               3/9/20              106
Sudan                   3/9/20                3
Suriname                3/9/20                8
Sweden                  3/9/20             3069
Switzerland             3/9/20            12928
Syria                   3/9/20                5
Taiwan*                 3/9/20              267
Tanzania                3/9/20               13
Thailand                3/9/20             1136
Timor-Leste             3/9/20                1
Togo                    3/9/20               25
Trinidad and Tobago     3/9/20               66
Tunisia                 3/9/20              227
Turkey                  3/9/20             5698
US                      3/9/20           101657
Uganda                  3/9/20               23
Ukraine                 3/9/20              310
United Arab Emirates    3/9/20              405
United Kingdom          3/9/20            14543
Uruguay                 3/9/20              238
Uzbekistan              3/9/20               88
Venezuela               3/9/20              107
Vietnam                 3/9/20              163
West Bank and Gaza      3/9/20               91
Zambia                  3/9/20               22
Zimbabwe                3/9/20                5

[176 rows x 2 columns]

If we have a look now at the 'US' row, we can see a value of 101657, which represents the value that we were expecting, as that is the value on 27/03, which should indicate the highest (max) number of cases.